src/neuron/container/memory_usage.hpp

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <cstdio>
+#include <vector>
+#include <ostream>
+#include <cmath>
+#include <sstream>
+
+namespace neuron::container {
+/** @brief Size and capacity in bytes. */
+struct VectorMemoryUsage {
+    VectorMemoryUsage() = default;
+    VectorMemoryUsage(size_t size, size_t capacity)
+        : size(size)
+        , capacity(capacity) {}
+
+    /** Compute the memory requirements of the `std::vector`.
+     *
+     * Note, this returns the size and capacity of the memory allocated by
+     * the `std::vector`. If the element allocate memory, that memory
+     * isn't included.
+     *
+     * Essentially,
+     *
+     *     size = vec.size() * sizeof(T);
+     *     capacity = vec.capacity() * sizeof(T);
+     */
+    template <class T, class A>
+    VectorMemoryUsage(const std::vector<T, A>& vec)
+        : size(vec.size() * sizeof(T))
+        , capacity(vec.capacity() * sizeof(T)) {}
+
+    /// @brief Number of bytes used.
+    size_t size{};
+
+    /// @brief Number of bytes allocated.
+    size_t capacity{};
+
+    inline const VectorMemoryUsage& operator+=(const VectorMemoryUsage& other) {
+        size += other.size;
+        capacity += other.capacity;
+
+        return *this;
+    }
+};
+
+
+/** @brief Memory usage of a storage/soa container. */
+struct StorageMemoryUsage {
+    /// @brief The memory usage of the heavy data in a soa.
+    VectorMemoryUsage heavy_data{};
+
+    /// @brief The memory usage for the stable identifiers in a soa.
+    VectorMemoryUsage stable_identifiers{};
+
+    inline const StorageMemoryUsage& operator+=(const StorageMemoryUsage& other) {
+        heavy_data += other.heavy_data;
+        stable_identifiers += other.stable_identifiers;
+
+        return *this;
+    }
+
+    inline size_t max_capacity() const {
```

Comment:
Where are these `max_capacity()` methods used?
(also, I believe `inline` is implicit here anyway)